app/state/services.py

```diff
@@ -137,30 +131,76 @@ def get_ip(self, headers: Mapping[str, str]) -> IPAddress:
         return ip
 
 
-def fetch_geoloc_db(ip: IPAddress) -> Geolocation:
-    """Fetch geolocation data based on ip (using local db)."""
-    assert geoloc_db is not None
+async def fetch_geoloc(
+    ip: IPAddress,
+    headers: Optional[Mapping[str, str]] = None,
+) -> Optional[Geolocation]:
+    """Fetch geolocation data based on ip."""
+    geoloc = fetch_geoloc_cloudflare(ip, headers)
 
-    res = geoloc_db.city(ip)
+    if geoloc is None:
+        geoloc = fetch_geoloc_nginx(ip, headers)
 
-    if res.country.iso_code is not None:
-        acronym = res.country.iso_code.lower()
-    else:
-        acronym = "XX"
+    if geoloc is None:
+        geoloc = await fetch_geoloc_web(ip)
+
+    return geoloc
+
+
+def fetch_geoloc_cloudflare(
+    ip: IPAddress,
+    headers: Mapping[str, str],
+) -> Optional[Geolocation]:
+    """Fetch geolocation data based on ip (using cloudflare headers)."""
+    if not all(
+        key in headers for key in ("CF-IPCountry", "CF-IPLatitude", "CF-IPLongitude")
+    ):
+        return
+
+    country_code = headers["CF-IPCountry"].lower()
+    latitude = float(headers["CF-IPLatitude"])
+    longitude = float(headers["CF-IPLongitude"])
 
     return {
-        "latitude": res.location.latitude or 0.0,
-        "longitude": res.location.longitude or 0.0,
+        "latitude": latitude,
+        "longitude": longitude,
         "country": {
-            "acronym": acronym,
-            "numeric": country_codes[acronym],
+            "acronym": country_code,
+            "numeric": country_codes[country_code],
+        },
+    }
+
+
+def fetch_geoloc_nginx(
+    ip: IPAddress,
+    headers: Mapping[str, str],
+) -> Optional[Geolocation]:
+    """Fetch geolocation data based on ip (using nginx headers)."""
+    if not all(
+        key in headers for key in ("X-Country-Code", "X-Latitude", "X-Longitude")
+    ):
+        return
+
+    country_code = headers["X-Country-Code"].lower()
+    latitude = float(headers["X-Latitude"])
+    longitude = float(headers["X-Longitude"])
+
+    return {
+        "latitude": latitude,
+        "longitude": longitude,
+        "country": {
+            "acronym": country_code,
+            "numeric": country_codes[country_code],
         },
     }
 
 
 async def fetch_geoloc_web(ip: IPAddress) -> Optional[Geolocation]:
     """Fetch geolocation data based on ip (using ip-api)."""
-    url = f"http://ip-api.com/line/{ip}"
+    if not ip.is_private:
+        url = f"http://ip-api.com/line/{ip}"
+    else:
+        url = "http://ip-api.com/line/"
```

Comment:
Noticed this too late - but we should actually be using `if not ip.is_loopback` here -- not `ip.is_private`

Reply:
Resolved here: https://github.com/osuAkatsuki/bancho.py/commit/112012a4a39c6cbb39f5a19a7523b084d70cc3cf